File: Leetcode/2461. Maximum Sum of Distinct Subarrays With Length K.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        freq = Counter(nums[:k])
        s = sum(nums[:k])
        res = 0
        n = len(nums)
        if len(freq) == k :
            res = max(res,s)
        for i in range(k,n):
            if nums[i] in freq:
                freq[nums[i]]+=1
            else:
                freq[nums[i]]=1
            s = s + nums[i] - nums[i-k]
            freq[nums[i-k]]-=1
            if freq[nums[i-k]]==0:
                del freq[nums[i-k]]
            if len(freq)==k:
                res = max(res,s)
        return res

File: Leetcode/2461. Maximum Sum of Distinct Subarrays With Length K.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        long long sum = 0, res = 0;
        int n = nums.size();
        
        for (int i = 0; i < k; ++i) {
            freq[nums[i]]++;
            sum += nums[i];
        }
        
        if (freq.size() == k) {
            res = max(res, sum);
        }
        
        for (int i = k; i < n; ++i) {
            freq[nums[i]]++;
            sum += nums[i] - nums[i - k];
            
            if (--freq[nums[i - k]] == 0) {
                freq.erase(nums[i - k]);
            }
            
            if (freq.size() == k) {
                res = max(res, sum);
            }
        }
        
        return res;
    }
};
